File: visualizer/pyvista/dive_stats.py
"""Dive statistics computation and formatting."""

import numpy as np
from flag_decoder import flag_severity


def compute_stats(data: dict) -> dict:
    """Compute dive statistics from parsed navigation data.

    Args:
        data: Dict from parse_nav_csv()

    Returns:
        Dict of statistics
    """
    positions = data['positions']
    timestamps = data['timestamps']
    flags = data['flags']
    delta_distances = data['delta_distances']

    duration_ms = timestamps[-1] - timestamps[0]
    duration_s = duration_ms / 1000.0

    total_distance = np.sum(delta_distances)

    depths = -positions[:, 2]
    max_depth = np.max(depths)

    avg_speed = total_distance / duration_s if duration_s > 0 else 0.0

    flagged_count = np.sum(flags != 0)
    flagged_percent = 100.0 * flagged_count / data['count']

    warning_count = sum(1 for f in flags if flag_severity(f) == 'warning')
    critical_count = sum(1 for f in flags if flag_severity(f) == 'critical')

    bbox_min = np.min(positions, axis=0)
    bbox_max = np.max(positions, axis=0)
    bbox_size = bbox_max - bbox_min

    return {
        'duration_s': duration_s,
        'duration_ms': duration_ms,
        'total_distance': total_distance,
        'max_depth': max_depth,
        'avg_speed': avg_speed,
        'sample_count': data['count'],
        'flagged_count': flagged_count,
        'flagged_percent': flagged_percent,
        'warning_count': warning_count,
        'critical_count': critical_count,
        'bbox_min': bbox_min,
        'bbox_max': bbox_max,
        'bbox_size': bbox_size,
    }


def print_stats(stats: dict) -> None:
    """Pretty-print dive statistics to console."""
    print("\n=== Dive Statistics ===")
    print(f"Duration:        {stats['duration_s']:>10.2f} s")
    print(f"Total Distance:  {stats['total_distance']:>10.2f} m")
    print(f"Max Depth:       {stats['max_depth']:>10.2f} m")
    print(f"Avg Speed:       {stats['avg_speed']:>10.4f} m/s")
    print(f"Sample Count:    {stats['sample_count']:>10d}")
    print(f"\n=== Data Quality ===")
    print(f"Flagged Samples: {stats['flagged_count']:>10d} ({stats['flagged_percent']:.1f}%)")
    print(f"  Warnings:      {stats['warning_count']:>10d}")
    print(f"  Critical:      {stats['critical_count']:>10d}")
    print(f"\n=== Bounding Box ===")
    print(f"Min:             ({stats['bbox_min'][0]:>7.2f}, {stats['bbox_min'][1]:>7.2f}, {stats['bbox_min'][2]:>7.2f})")
    print(f"Max:             ({stats['bbox_max'][0]:>7.2f}, {stats['bbox_max'][1]:>7.2f}, {stats['bbox_max'][2]:>7.2f})")
    print(f"Size:            ({stats['bbox_size'][0]:>7.2f}, {stats['bbox_size'][1]:>7.2f}, {stats['bbox_size'][2]:>7.2f})")
    print()

File: visualizer/web/js/dive-stats.js
import { convertValue, getUnit, getPrecision, cycleUnit } from './units.js';

export function computeStats(parsedData) {
  const { timestamps, positions, deltaDistances, flags, count } = parsedData;

  if (count === 0) {
    throw new Error('No data points to compute stats');
  }

  const durationMs = timestamps[count - 1] - timestamps[0];
  const duration = durationMs / 1000.0;

  let totalDistance = 0;
  for (let i = 0; i < count; i++) {
    totalDistance += deltaDistances[i];
  }

  const minPos = [Infinity, Infinity, Infinity];
  const maxPos = [-Infinity, -Infinity, -Infinity];

  let maxDepth = 0;

  for (let i = 0; i < count; i++) {
    const px = positions[i * 3 + 0];
    const py = positions[i * 3 + 1];
    const pz = positions[i * 3 + 2];

    const vizX = px;
    const vizY = pz;
    const vizZ = py;

    minPos[0] = Math.min(minPos[0], vizX);
    minPos[1] = Math.min(minPos[1], vizY);
    minPos[2] = Math.min(minPos[2], vizZ);

    maxPos[0] = Math.max(maxPos[0], vizX);
    maxPos[1] = Math.max(maxPos[1], vizY);
    maxPos[2] = Math.max(maxPos[2], vizZ);

    maxDepth = Math.max(maxDepth, Math.abs(pz));
  }

  // Compute depth percentiles for color normalization
  const depths = new Float32Array(count);
  for (let i = 0; i < count; i++) {
    depths[i] = Math.abs(positions[i * 3 + 2]);
  }
  depths.sort();
  const depthP5 = depths[Math.floor(count * 0.05)] ?? 0;
  const depthP95 = depths[Math.floor(count * 0.95)] ?? maxDepth;

  const avgSpeed = duration > 0 ? totalDistance / duration : 0;

  let flaggedCount = 0;
  for (let i = 0; i < count; i++) {
    if (flags[i] !== 0x00) {
      flaggedCount++;
    }
  }
  const flaggedPercent = count > 0 ? (flaggedCount / count) * 100 : 0;

  return {
    duration,
    totalDistance,
    maxDepth,
    depthP5,
    depthP95,
    avgSpeed,
    flaggedPercent,
    boundingBox: {
      min: minPos,
      max: maxPos
    },
    sampleCount: count
  };
}

export function formatStats(stats) {
  const {
    duration,
    totalDistance,
    maxDepth,
    avgSpeed,
    flaggedPercent,
    sampleCount
  } = stats;

  const formatDuration = (seconds) => {
    const totalSeconds = Math.floor(seconds);
    const mins = Math.floor(totalSeconds / 60);
    const secs = totalSeconds % 60;
    return `${mins}:${secs.toString().padStart(2, '0')}`;
  };

  const flaggedClass = flaggedPercent > 50 ? 'critical' : flaggedPercent > 10 ? 'warning' : '';

  const distanceValue = convertValue('distance', totalDistance);
  const distanceUnit = getUnit('distance');
  const distancePrecision = getPrecision('distance');

  const depthValue = convertValue('depth', maxDepth);
  const depthUnit = getUnit('depth');
  const depthPrecision = getPrecision('depth');

  const speedValue = convertValue('speed', avgSpeed);
  const speedUnit = getUnit('speed');
  const speedPrecision = getPrecision('speed');

  return `
    <div class="stat-row">
      <span class="stat-label">Duration:</span>
      <span class="stat-value">${formatDuration(duration)}</span>
    </div>
    <div class="stat-row">
      <span class="stat-label">Distance:</span>
      <span class="stat-value"><span class="stat-number" data-stat="distance">${distanceValue.toFixed(distancePrecision)}</span> <span class="stat-unit" data-stat="distance">${distanceUnit}</span></span>
    </div>
    <div class="stat-row">
      <span class="stat-label">Max Depth:</span>
      <span class="stat-value"><span class="stat-number" data-stat="depth">${depthValue.toFixed(depthPrecision)}</span> <span class="stat-unit" data-stat="depth">${depthUnit}</span></span>
    </div>
    <div class="stat-row">
      <span class="stat-label">Avg Speed:</span>
      <span class="stat-value"><span class="stat-number" data-stat="speed">${speedValue.toFixed(speedPrecision)}</span> <span class="stat-unit" data-stat="speed">${speedUnit}</span></span>
    </div>
    <div class="stat-row">
      <span class="stat-label">Flagged:</span>
      <span class="stat-value ${flaggedClass}">${flaggedPercent.toFixed(1)}%</span>
    </div>
    <div class="stat-row">
      <span class="stat-label">Samples:</span>
      <span class="stat-value">${sampleCount.toLocaleString()}</span>
    </div>
  `;
}

export function bindUnitToggles(stats, container) {
  const RAW_VALUES = {
    distance: stats.totalDistance,
    depth: stats.maxDepth,
    speed: stats.avgSpeed,
  };

  const unitElements = container.querySelectorAll('.stat-unit');

  unitElements.forEach(element => {
    element.addEventListener('click', () => {
      const statName = element.getAttribute('data-stat');
      if (!statName || !RAW_VALUES.hasOwnProperty(statName)) {
        return;
      }

      cycleUnit(statName);

      const rawValue = RAW_VALUES[statName];
      const convertedValue = convertValue(statName, rawValue);
      const precision = getPrecision(statName);
      const unit = getUnit(statName);

      const numberElement = container.querySelector(`.stat-number[data-stat="${statName}"]`);
      const unitElement = container.querySelector(`.stat-unit[data-stat="${statName}"]`);

      if (numberElement) {
        numberElement.textContent = convertedValue.toFixed(precision);
      }
      if (unitElement) {
        unitElement.textContent = unit;
      }
    });
  });
}
